Generate an .sframe section with a skeleton header

Sterling-Augustine opened the pull request:
This continues the sframe implementation discussed previously.

Of note, this also adds some target dependent functions to the object file. Additional fields will be needed later. It would be possible to do all of this inside the sframe implementation itself if it feels a little messy and specialized, but generally I think that target info goes with target info.

Another question is if we want a sentinel value for unimplemented sframe abi arches, or a std::optional. Both work.

Sterling-Augustine committed:
Generate an .sframe section with a skeleton header

This continues the sframe implementation discussed previously.

Of note, this also adds some target dependent functions to the object
file. Additional fields will be needed later. It would be possible to
do all of this inside the sframe implementation itself if it feels
a little messy and specialized, but generally I think that target info
goes with target info.

Sterling-Augustine committed:
Fix formatting

labath commented on llvm/lib/MC/MCSFrame.cpp:
```diff
@@ -0,0 +1,94 @@
+
+  void EmitPreamble() {
+    Streamer.emitInt16(Magic);
+    Streamer.emitInt8(static_cast<uint64_t>(Version::V2));
```
I'd use uint8_t as that's the backing type of the Version enum, even though emitInt8 takes a uint64_t. Casting to uint64_t looks very odd here.

labath reviewed (commented):
> Another question is if we want a sentinel value for unimplemented sframe abi arches, or a std::optional.

For me, that depends on whether the value is going to end up in the final object file. If yes, then it should definitely be an enumerator. If not, then I'd slightly prefer not putting it there, either via std::optional, using `{}` instead of `ABI::Undefined`, or by moving the logic into a helper function in MCSFrame (where the unimplemented cases could presumably be handled by `llvm_unreachable`).

I don't see anything here or in #149935 that would prevent the sframe section from being generated for an unsupported target. Is that intentional? In my experiments, the GNU toolchain ignored the `--sframe` argument on an unsupported target -- which isn't particularly helpful (I would expect at least a warning), but I guess it's better than emitting something that can't be used.

Sterling-Augustine committed:
Address comments:

Switch to std::optional for SFrameABI
Use uint8_t for casting to MCObjectStreamer

Sterling-Augustine commented:
Anyone else?

jh7370 commented on llvm/test/MC/ELF/cfi-sframe.s:
```diff
@@ -0,0 +1,26 @@
+.cfi_sections .sframe
+	
+f1:
+        .cfi_startproc
+        nop
+        .cfi_endproc
```
Nit: excessive indentation?

Sterling-Augustine replied:
This matches the other cfi tests. A lot of the code inside MC is somewhat old, so there is the tension between matching what currently exists and updating.

Sterling-Augustine commented on llvm/include/llvm/MC/MCSFrame.h:
```diff
@@ -0,0 +1,36 @@
+//===- MCSFrame.h - Machine Code SFrame support -----------------*- C++ -*-===//
```
Done

Sterling-Augustine commented on llvm/lib/MC/MCSFrame.cpp:
```diff
@@ -0,0 +1,98 @@
+  SFrameEmitterImpl Emitter(Streamer);
+
+  MCSection *Section = Context.getObjectFileInfo()->getSFrameSection();
+  // Not strictly necessary, but gas always aligns to 8, so match that.
```
For a brand-new feature, matching our implementation against the reference implementation makes it way easier to find bugs. Especially because the gnu sframe dumper omits a lot of details that our dumper doesn't yet support either. After things mature a little bit, maybe we can reconsider.

Sterling-Augustine committed:
Address comments.

Fix formatting

Sterling-Augustine commented:
I think I have addressed all the comments. Anyone want to approve?

jh7370 commented on llvm/lib/MC/MCSFrame.cpp:
```diff
@@ -0,0 +1,98 @@
+    FRESubSectionEnd = Streamer.getContext().createTempSymbol();
+  }
+
+  void EmitPreamble() {
```
Not sure how I missed this in the first pass, but all these function names should use camelCase (i.e. `emitPreamble` here, `emitHeader` etc), per LLVM naming policy. It's a brand new file, so there's no need to match any existing style otherwise.

jh7370 reviewed (approved):
LGTM, except that the function names use the wrong capitalization style.

labath reviewed (approved):
LGTM, modulo James's comments.

Sterling-Augustine committed:
Address comments

Sterling-Augustine committed:
Merge branch 'main' into sf_header

Sterling-Augustine committed:
Remove currently unused includes
